Archive and compare iterations.log for implicit coupling (#743)
precice/tutorials: tools/tests/README.md

Reviewer: When merging `develop` into this branch, you will find a new section on "Numerical regressions": https://precice.org/dev-docs-system-tests.html#numerical-regressions

We should list these files there as well.

Author: yeah done

3. `system-tests-compare.log`: The logs for the comparison to the reference results.

In addition, in the directories of the cases executed, you can find `system-tests-<case>.log` files.

### Numerical regressions

When the tests fail at the results comparison step, this typically means that there are numerical regressions (unless something wrong went undetected in a previous step). We use [fieldcompare](https://gitlab.com/dglaeser/fieldcompare) to compare all [preCICE exports](https://precice.org/configuration-export.html) to reference results generated from a previous run. Relevant files:

- `precice-exports/`: The coupling meshes of the test run.
- `reference-results/`: The coupling meshes of the reference run, as stored on Git LFS, expanded into `reference-results-unpacked`.
- `diff-results/`: Numerical difference of the results in the two directories (computed with `fieldcompare dir --diff precice-exports/ reference/`). These are only present on failed comparisons.
- `iterations-logs/`: For implicit-coupling runs, archived `precice-*-iterations.log` files from the test run, compared by SHA-256 hash against reference copies.

> - `*.iterations-logs/`: Reference iterations logs stored next to each reference `.tar.gz` on Git LFS (e.g. `fluid-openfoam_solid-openfoam.iterations-logs/`).

Reviewer: These lines can be simplified a bit (but I don't currently have the option to directly suggest changes):

```
- `iterations-logs/`: The `precice-*-iterations.log` files of the test run. Only present in implicit coupling tests and compared by SHA-256 hash against reference copies.
- `*.iterations-logs/`: The reference iterations log files, stored next to each reference results archive on Git LFS (e.g. `fluid-openfoam_solid-openfoam.iterations-logs/`).
```

Author: fine i have done it as per your wording

2. Create label as a trigger for workflow (under `Issues`->`Labels`->`New label`).
3. Give permissions: In the [preCICE organization settings](https://github.com/organizations/precice/settings), you need to add the new repository to the action secret `WORKFLOW_DISPATCH_TOKEN` and to the default actions runner group. Adding the new label directly to the pull request in which you add the workflow should already trigger the system tests.

## Implementation details

Each tutorial contains automation scripts (mainly `run.sh` and `clean.sh`), as well as metadata (`metadata.yaml`). The metadata file describes the available cases, how to run them, as well as their dependencies. A central `tests.yaml` file defines test suites, which execute different combinations of cases. The Python script `systemtests.py` executes the tests, filter for specific test suites.

Read more about the system tests in the publication [System Regression Tests for the preCICE Coupling Ecosystem](https://doi.org/10.14279/eceasst.v83.2614).

[![ECEASST](https://img.shields.io/badge/DOI-10.14279%2Feceasst.v83.2614-green)](https://doi.org/10.14279/eceasst.v83.2614)

> <details markdown="1"><summary>Expand the implementation details...</summary>

Reviewer: Something went wrong in the conflict resolution here.

Author: checked against develop , that section matches ; only the two iterations bullets i added.

Reviewer: Ah, indeed, the diff was a bit confusing.

When the tests fail at the results comparison step, this typically means that there are numerical regressions (unless something wrong went undetected in a previous step). We use [fieldcompare](https://gitlab.com/dglaeser/fieldcompare) to compare all [preCICE exports](https://precice.org/configuration-export.html) to reference results generated from a previous run. Relevant files:

- `precice-exports/`: The coupling meshes of the test run.

> - `reference-results/`: The coupling meshes of the reference run, as stored on Git LFS, expanded into `reference-results-unpacked`. For implicit-coupling tests, the reference `.tar.gz` also contains a separate `{case}.iterations-logs/` directory with the reference `precice-*-iterations.log` files.

Reviewer: ```suggestion
- `reference-results/`: The coupling meshes of the reference run, as stored on Git LFS, expanded into `reference-results-unpacked`. For test cases using implicit coupling, the reference `.tar.gz` also contains the reference `precice-*-iterations.log` files.
```

- `diff-results/`: Numerical difference of the results in the two directories (computed with `fieldcompare dir --diff precice-exports/ reference/`). These are only present on failed comparisons.

> - `iterations-logs/`: The `precice-*-iterations.log` files of the test run. Only present in implicit coupling tests and compared by SHA-256 hash against the copies unpacked from `{case}.iterations-logs/` in the reference `.tar.gz`.

Reviewer: ```suggestion
- `iterations-logs/`: The `precice-*-iterations.log` files of the test run. Only present in test cases using implicit coupling. The comparisons to references only take into account the file SHA-256 checksums.
```
